modules/objdetect/src/qrcode.cpp

```diff
@@ -3816,33 +3835,467 @@ bool QRDetectMulti::computeTransformationPoints(const size_t cur_ind)
+           }
+
+    vector<Point2f> getQRCorners() const {
+        Point2f a1 = rightPattern.getQRCorner().second;
+        Point2f a2 = rightPattern.getCornerForIntersection().second;
+
+        Point2f b1 = bottomPattern.getQRCorner().second;
+        Point2f b2 = bottomPattern.getCornerForIntersection().second;
+
+        Point2f rightBottom = intersectionLines(a1, a2, b1, b2);
+
+        return {centerPattern.getQRCorner().second, rightPattern.getQRCorner().second, rightBottom, bottomPattern.getQRCorner().second};
+    }
+
+    static QRCode checkCompatibilityPattern(const FinderPatternInfo &_pattern1, const FinderPatternInfo& _pattern2, const FinderPatternInfo& _pattern3,
+                                            Point3i& index, QRDetectorParameters& qrDetectorParameters) {
+        FinderPatternInfo pattern1 = _pattern1, pattern2 = _pattern2, pattern3 = _pattern3;
+        Point2f centerQR;
+        float distance = std::numeric_limits<float>::max();
+
+        if (abs(pattern1.minQrAngle - pattern2.minQrAngle) >  qrDetectorParameters.maxRotateDiff ||
+            abs(pattern1.minQrAngle - pattern3.minQrAngle) > qrDetectorParameters.maxRotateDiff) // check maxRotateDiff
+            return QRCode(pattern1, pattern2, pattern3, centerQR, distance);
+        if (max(pattern1.moduleSize, pattern2.moduleSize) / min(pattern1.moduleSize, pattern2.moduleSize) > qrDetectorParameters.maxRelativeModuleDiff ||
+            max(pattern1.moduleSize, pattern3.moduleSize) / min(pattern1.moduleSize, pattern3.moduleSize) > qrDetectorParameters.maxRelativeModuleDiff)
+            return QRCode(pattern1, pattern2, pattern3, centerQR, distance);
+        // QR code:
+        // center       right
+        //    1 ________ 2
+        //     |_|    |_|
+        //     |     /  |
+        //     |    /   |
+        //     |   /    |
+        //     |_ /     |
+        //     |_|______|
+        //    4
+        //  bottom
+
+        // sides length check
+        const float side1 = sqrt(normL2Sqr<float>(pattern1.center - pattern2.center));
+        const float side2 = sqrt(normL2Sqr<float>(pattern1.center - pattern3.center));
+        const float side3 = sqrt(normL2Sqr<float>(pattern2.center - pattern3.center));
+        std::array<float, 3> sides = {side1, side2, side3};
+        std::sort(sides.begin(), sides.end());
+        // check sides diff
+        if (sides[1] / sides[0] < qrDetectorParameters.maxRelativeModuleDiff) {
+            // find center pattern
+            if (side1 > side2 && side1 > side3) { // centerPattern is pattern3
+                std::swap(pattern3, pattern1); // now pattern1 is centerPattern
+                std::swap(index.x, index.z);
+            }
+            else if (side2 > side1 && side2 > side3) { // centerPattern is pattern2
+                std::swap(pattern2, pattern1); // now pattern1 is centerPattern
+                std::swap(index.x, index.y);
+            }
```

Comment:
The code with sort and then swaps looks very cryptic. Not clear, what if `side2 == side1`. I propose better solution:
1. Define a struct with length and reference(!) to FinderPatternInfo.
2. Use std::array of the structures and sort it by length.
3. Use array[0] as center pattern everywhere. All other - left and right.

Reply:
Discussed offline. Not relevant.